### ISSUE TITLE
✨ Incorporation of a Configurable Parameter to Specify Address, Name, and Password of the Registry from which the Dagger Engine Image is Downloaded

### ISSUE THREAD

**Issue by wamoina** [dagger/dagger#5417]
### What are you trying to do?

In the context of our project, we are attempting to set up Dagger, a task runner and workflow orchestrator, in an environment where all services and resources are exclusively accessible through our intranet. To this end, we need to download the Dagger engine image from a private Docker or Podman registry hosted within our intranet.

Currently, Dagger's code is structured in such a way that the registry's domain (in this case, dl.dagger.io) is hardcoded as a non-configurable global variable in the download.py file. For instance:
```
...
## omitted code
DEFAULT_CLI_HOST = "dl.dagger.io"

...
## omitted code

class Downloader:
    """Download the dagger CLI binary."""

    CLI_BIN_PREFIX = "dagger-"
    CLI_BASE_URL = f"https://{DEFAULT_CLI_HOST}/dagger/releases"
```

### Why is this important to you?

This feature is critical for us because all our services and resources are exclusively accessible through our private intranet. This network restriction means we need to establish parameters for a private repository from which the Dagger engine image can be downloaded, which is currently not possible due to hard-coded values in the software.

### How are you currently working around this?

For now, we do not have a satisfactory solution. Our current procedure involves opening access to the internet, which is far from ideal due to the need to maintain a separate Dagger code base and merge changes from the main project. This method is error prone and does not scale well with project growth.

We believe the addition of this feature could benefit others in similar environments or anyone who needs to download the Dagger engine image from a private repository for security or other reasons.

**Comment by nipuna-perera** [comment]
The request here is important for my company too. We proxy all our docker images via our internal docker proxy (Artifactory). I can set `_EXPERIMENTAL_DAGGER_REGISTRY` and download via the proxy. I have to download the ghcr.io image as dagger.io is not an allowlisted registry in our proxy. However, the problem is when I want to distribute this to other developers within the company. 

If I set the experimental flag, I am hardcoding the Dagger version in the user space. Then the developer has to mind/care about the specific dagger version and set the correct one whenever we upgrade the dependencies of our SDK. It's not very scalable like that. It would be great if we could programatically set the dagger registry so when we upgrade our SDK to a new dagger version, all the consumers get it automatically.

**Comment by mircubed** [comment]
@samalba @gerhard to review for resourcing

**Comment by gerhard** [comment]
Congratulations @wamoina for your first GitHub activity - this issue!

TL;DR: you can use a custom registry today via `_EXPERIMENTAL_DAGGER_RUNNER_HOST=docker-image://<YOUR_PRIVATE_REGISTRY>...`

---
There are a few parts to how this currently works:
1. Compatible CLI version is hard-coded in each SDK
2. CLI download host defaults to `dl.dagger.io`
3. Engine image registry defaults to `registry.dagger.io`

The hard-coded CLI version in the SDK will continue behaving the same. In other words, the version that the SDK depends on should always be available at `DAGGER_CLI_DOWNLOAD_HOST`.

On the `CLI_DOWNLOAD_HOST`, the following paths should be resolvable:
- `/dagger/releases/<VERSION>/dagger_v<VERSION>_<OS>_<ARCH>.<EXT>`
- `/<VERSION>/checksums.txt`

> **Note**
> Artefacts that  need to be available: https://github.com/dagger/dagger/releases/tag/v0.6.4

The Engine image registry is easier to handle, since `DAGGER_ENGINE_REGISTRY` would need to serve `/engine:<VERSION>`.

Before we go further with this, are you sure that hosting these artefacts, and updating when new ones become available, is what you want to do? This will require new artefacts to be made available on your intranet before you can use them.

Maybe vendoring the CLI binary, and using e.g. `_EXPERIMENTAL_DAGGER_CLI_BIN=./bin/dagger_<VERSION>` together with `_EXPERIMENTAL_DAGGER_RUNNER_HOST=docker-image://<YOUR_PRIVATE_REGISTRY>/engine:<VERSION>` would serve the purpose.

WDYT?

**Comment by helderco** [comment]
@wamoina it looks like you're running with `python main.py` and thus relying on the automatic provisioning from the SDK but that’s just a convenience to facilitate onboarding. You can get and install the `dagger` CLI however’s best for your use case and then use `dagger run python main.py` which provides a running session from `dagger run` to Python and makes the SDK skip the convenience of downloading the CLI for you.

The CLI, as you know, will then try to download the engine container from its default location but you can change where it’s pulled from. See @gerhard’s [comment above](https://github.com/dagger/dagger/issues/5417#issuecomment-1658290424).

Seeing as this is a documentation issue, I’ve created a new one to specifically target the *offline* use case and closing this one as a duplicate.

Let us know if there’s any specifics to your setup that you’re not sure how to handle. You can follow up in:
- https://github.com/dagger/dagger/issues/6275
- https://github.com/dagger/dagger/issues/5135
- https://github.com/dagger/dagger/issues/5240
- https://github.com/dagger/dagger/issues/4217
- https://github.com/dagger/dagger/issues/6077

There's some details in the [operator manual](https://github.com/dagger/dagger/blob/main/core/docs/d7yxc-operator_manual.md#faq) as well.
